airbyte-integrations/bases/base-python/base_python/sdk/streams/http.py

```diff
@@ -0,0 +1,251 @@
+import copy
+from abc import ABC, abstractmethod
+from typing import Dict, Iterable, Any, Optional
+
+import requests
+
+from .auth.core import HttpAuthenticator, NoAuth
+from .exceptions import DefaultBackoffException, CustomBackoffException
+from .rate_limiting import custom_backoff_handler, default_backoff_handler
+
+from .core import Stream
+
+
+class HttpStream(Stream, ABC):
+
+    def __init__(self, authenticator: HttpAuthenticator = NoAuth(), parent_stream: 'HttpStream' = None):
+        self._authenticator = authenticator
+        self._parent_stream = parent_stream
+        self._session = requests.Session()
+
+    @property
+    @abstractmethod
+    def url_base(self) -> str:
+        """
+        :return: URL base for the  API endpoint e.g: if you wanted to hit https://myapi.com/v1/some_entity then this should return "https://myapi.com/v1/"
+        """
+
+    @property
+    def http_method(self) -> str:
+        """
+        Override if needed. See get_request_data if using POST.
+        """
+        return "GET"
+
+    @property
+    def authenticator(self) -> HttpAuthenticator:
+        return self._authenticator
+
+    def get_request_configurations(self, stream_state: Dict[str, Any], parent_stream_record: Dict = None) -> Iterable[Optional[Dict]]:
+        """
+        Override this method to control the number of HTTP requests (not counting pagination) this stream makes.
+
+        For example, if a single request request reads data for a particular date and you want to read data from multiple dates, then this method should
+        return one dict for each request that should be made e.g: [{'date':'01-01-2020'}, {'date':'01-02-2020'}], etc.
+
+        Alternatively, if you want to make no requests (e.g: if you don't want to use up more than 30% of your API's rate limit for the day) then
+        return an empty iterable.
+
+        Each element in the returned iterable will be passed as input to most other methods in this class to initiate a single HTTP request, plus any
+        subsequent requests needed for pagination.
+
+        :return: An iterable (list or generator) of request configurations
+        """
+        return [None]
+
+    @abstractmethod
+    def get_next_page_token(self, decoded_response: Dict[str, Any]) -> Optional[Dict[str, Any]]:
+        """
+        Override this method to define a pagination strategy.
+
+        The value returned from this method is passed to most other methods in this class. Use it to form a request e.g: set headers or query params.
+
+        :return: The token for the next page from the input response object. Returning None means there are no more pages to read in this response.
+        """
+
+    @abstractmethod
+    def path(
```

Comment:
is there any real use case when path depend on any of these parameters?
maybe just `property` is enough?

Reply:
parent streams